File: automedal/agent/phases/experimenter_eval.py
"""Experimenter-eval phase — parse training result, write journal, commit/revert."""

from __future__ import annotations

from automedal.agent.events import EventSink
from automedal.agent.kernel import RunReport
from automedal.agent.phases._common import run_phase
from automedal.agent.tools import (
    READ_FILE, WRITE_FILE, EDIT_FILE, LIST_DIR, GREP, RUN_SHELL,
)


def _eval_tools():
    return [READ_FILE, WRITE_FILE, EDIT_FILE, LIST_DIR, GREP, RUN_SHELL]


async def run(
    *,
    provider,
    events: EventSink | None,
    exp_id: str,
    best_val_loss: float | str,
    train_rc: int,
    final_loss: float | str,
    max_steps: int = 30,
) -> RunReport:
    return await run_phase(
        phase="experimenter_eval",
        provider=provider,
        tools=_eval_tools(),
        events=events,
        max_steps=max_steps,
        slots={
            "exp_id": exp_id,
            "best_val_loss": best_val_loss,
            "train_rc": train_rc,
            "final_loss": final_loss,
        },
    )

File: internal/agent/phases/experimenter_eval.go
// Experimenter-eval phase — parse training result, write journal, commit/revert.
package phases

import (
	"context"

	"github.com/Flameingmoy/automedal/internal/agent"
	"github.com/Flameingmoy/automedal/internal/agent/tools"
)

// ExperimenterEvalArgs is the slot dict for the experimenter-eval phase.
type ExperimenterEvalArgs struct {
	ExpID       string
	BestValLoss any
	TrainRC     int
	FinalLoss   any
	MaxSteps    int
}

// RunExperimenterEval runs the experimenter-eval phase.
func RunExperimenterEval(ctx context.Context, chat agent.ChatStreamFunc, events *agent.EventSink, a ExperimenterEvalArgs) (agent.RunReport, error) {
	base := []tools.Tool{
		tools.ReadFile, tools.WriteFile, tools.EditFile, tools.ListDir, tools.Grep, tools.RunShell,
	}
	maxSteps := a.MaxSteps
	if maxSteps <= 0 {
		maxSteps = 30
	}
	return RunPhase(ctx, RunPhaseConfig{
		Phase:    "experimenter_eval",
		Chat:     chat,
		Tools:    base,
		Events:   events,
		MaxSteps: maxSteps,
		Slots: map[string]any{
			"exp_id":        a.ExpID,
			"best_val_loss": a.BestValLoss,
			"train_rc":      a.TrainRC,
			"final_loss":    a.FinalLoss,
		},
	})
}
